mobile-browser-based-version/src/logic/communication/client.ts

```diff
@@ -1,57 +1,53 @@
 import * as tf from '@tensorflow/tfjs'
+import { Task } from '../task_definition/base/task'
+import { TrainingInformant } from '../training/training_informant'
 
-export class Client {
-  serverURL: any;
-  task: any;
-  constructor (serverURL, task) {
+export abstract class Client {
+  serverURL: string;
+  task: Task;
+  constructor (serverURL: string, task: Task) {
     this.serverURL = serverURL
     this.task = task
   }
 
+  /**
+   * Init holder for federated, we need to fetch latest server round
+   * and model before training.
+   */
+  async updateModelAndRoundFromServer () {}
+
   /**
    * Handles the connection process from the client to any sort of
    * centralized server.
    */
-  async connect (epochs?): Promise<any> {
-    throw new Error('Abstract method')
-  }
+  abstract connect (epochs?:number): Promise<any>
 
   /**
    * Handles the disconnection process of the client from any sort
    * of centralized server.
    */
-  disconnect () {
-    throw new Error('Abstract method')
-  }
-
-  async onTrainBeginCommunication (model, trainingInformant) {
-
-  }
+  abstract disconnect (): Promise<any>
 
   /**
    * The training manager matches this function with the training loop's
    * onTrainEnd callback when training a TFJS model object. See the
    * training manager for more details.
    */
-  async onTrainEndCommunication (model, trainingInformant) {
-    trainingInformant.addMessage('Training finished.')
-  }
+  abstract onTrainEndCommunication (model, trainingInformant: TrainingInformant): Promise<void>
```

Comment:
I think the callback methods should not be abstract but rather be some sort of `default` methods, since the `Client` abstract class may include some routine common to every subclasses, such as displaying the RAM usage

Reply:
The RAM usage should be displayed inside the onRoundEnd call backs in the trainer? If there are no common subroutines, is it not better to keep it abstract?